packages/ics721/src/ibc_packet_receive.rs

```diff
@@ -232,19 +226,11 @@ fn create_callback_msg(
             load_nft_contract_for_class_id(deps.storage, local_class_id.to_string())
                 .map_err(|_| ContractError::NoNftContractForClassId(local_class_id.to_string()))
         } else {
-            // If its a creation action, we can use the instantiate2 function to get the nft contract
-            // we don't care of the contract is instantiated yet or not, as later submessage will instantiate it if its not.
-            // The reason we use instantiate2 here is because we don't know if it was already instantiated or not.
-
-            let cw721_code_id = CW721_CODE_ID.load(deps.storage)?;
-            // for creating a predictable nft contract using, using instantiate2, we need: checksum, creator, and salt:
-            // - using class id as salt for instantiating nft contract guarantees a) predictable address and b) uniqueness
-            // for this salt must be of length 32 bytes, so we use sha256 to hash class id
-            let mut hasher = Sha256::new();
-            hasher.update(local_class_id.as_bytes());
-            let salt = hasher.finalize().to_vec();
-
-            get_instantiate2_address(deps, env.contract.address.as_str(), &salt, cw721_code_id)
+            // here we know NFT contract is always instantiated, so it is safe getting it from storage
+            // NEVER use instantiate2 to get the NFT contract address here, since code id may change!
+            Ok(CLASS_ID_AND_NFT_CONTRACT_INFO
+                .load(deps.storage, &local_class_id)?
+                .address)
```

Comment:
This is the bug fix. Before it used `instantiate2` for retrieving nft contract, but in case cw721 code id is changing, it returns an non-existent, unknown address.

Code before:

```rust
            let cw721_code_id = CW721_CODE_ID.load(deps.storage)?;
            // for creating a predictable nft contract using, using instantiate2, we need: checksum, creator, and salt:
            // - using class id as salt for instantiating nft contract guarantees a) predictable address and b) uniqueness
            // for this salt must be of length 32 bytes, so we use sha256 to hash class id
            let mut hasher = Sha256::new();
            hasher.update(local_class_id.as_bytes());
            let salt = hasher.finalize().to_vec();

            get_instantiate2_address(deps, env.contract.address.as_str(), &salt, cw721_code_id)
```

Reply:
Why did we use salt to get address before? How to guarantee contract will always exist on retrieval?

Reply:
Callback happens after collection has been instantiated. ICS721 instantiates it and stores it in `CLASS_ID_AND_NFT_CONTRACT_INFO`.

salt is required and part of `instantiate2`. Using class id as salt guarantees that address always results in same escrowed nft contract (predictable). This is used durint instantiation of contract, which then stores predictable address in `CLASS_ID_AND_NFT_CONTRACT_INFO`. `CLASS_ID_AND_NFT_CONTRACT_INFO` is then used by mint and callbacks.

Reply:
Ah yeah so for instantiate2 we need the salt, but here we are now using `local_class_id` to retrieve from `CLASS_ID_AND_NFT_CONTRACT_INFO`. Just wondering where do we get `local_class_Id` from? Seems to be passed into the function.

Reply:
class_id is passed from source to target chain (e.g. nft contract `stars1xyz`) as part pf , then target chain attaches destination port and channel e.g. `wasm.osmo1ics721/channel-xyz/stars1xyz`). This is stored as key and nft contract as value in `CLASS_ID_AND_NFT_CONTRACT_INFO` map.